[WIP] layout: css-text-overflow

d-desiatkin opened the pull request:
<!-- Please describe your changes on the following line: -->
This PR aims to introduce new CSS tag support for layout 2020.

Stylo change is required. PR for stylo is:
https://github.com/servo/stylo/pull/122

Also some additional PR is needed to fix documentation generation on windows:
https://github.com/servo/stylo/pull/123

Implementation details:
Current plan is the following:
- [X] 1) Add ellipsis string layout to InlineFormattingContext creation. As a result we should have separate storage vector for ellipsis elements.
- [X] 2) Introduce that additional Fragments into Fragment tree.
- [X] 3 a.) Temporary place fragments in any stage that will be convenient to check that all shaping relation operations work fine. 
- [X] 3 b.) Modifiy fragments placement on Stacking context creation stage;
- [] 4) Fix placements bugs. Need to propperly extract baselines from box and text fragments 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix 

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

d-desiatkin committed:
[feature](text-overflow-ellipsis): Creation of storage for necesarry
parameters to alter rendering.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): Create separate fragments storage for ellipsis
elements

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): Try to reuse parts of line layout code for
ellipsis fragments positioning.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): Additional ellipsis fragments is rendered.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): MVP version

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[chore](text-overflow): Fix linter errors; Style improvements

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[bugfix](text-overflow): Clear line items before resetting cuttent line
segment. Part of the process of ellipsis Line Items layout

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

Loirooriol reviewed (commented):
It says "WIP", so I guess this should still be a draft. When you open a PR which is still not ready to be reviewed, please open it as a draft to avoid automatic review requests.

d-desiatkin commented:
@Loirooriol right now it is draft but I want to ask you several questions regarding the realization of the feature. I see that you had some conversations before
https://github.com/servo/servo/issues/13702
https://github.com/servo/servo/issues/17009
https://github.com/servo/servo/issues/15144

I have several questions on how you see this.  My current realization is already partially functional. At least 2 additional tests could already pass. Main problem however that I still do fragment positioning on line layout stage. I want to transfer it to stacking context and it seems to be doable. I mean the only information that is really required is already exist on stacking context creation stage. Although, I still want to position fragments using logical coordinates. And we already resolve their position in `components\layout_2020\flow\inline\line.rs` file. Is there some specific reason why we are doing it here? Can we postpone this operation? If we must do it like we do it know will double conversion look good for you? logical->physical->logical->physical?

d-desiatkin commented:
@Loirooriol  Another thing that I want you to look at, is introduction of ouroboros crate to handle the lifetime bounding of owned Strings in InlineFormattingContext and EllipsisStorage. That crate is very non idiomatic for rust, moreover it potentially hides unsafe code from servo codebase. That is what I personally really dislike. But I can not come up with nice solution to this, cause a lot of additional covariates must be added to IFC to make it work, and it makes the code kind of mess. What will be idiomatic Rust solution to the problem? Do you have a link on some educational material?

d-desiatkin commented on components/layout_2020/flow/inline/text_run.rs:
```diff
@@ -62,6 +72,120 @@ enum SegmentStartSoftWrapPolicy {
+// someone will suggest how to do it without turning InlineFormattingContext into mess with lots
+// of different lifetimes...
+
+#[ouroboros::self_referencing]
+#[derive(Debug)]
+pub(crate) struct BidiTextStorage {
+    pub text: String,
+    #[borrows(text)]
+    #[covariant]
+    pub bidi_info: BidiInfo<'this>,
+}
```
@Loirooriol Do you know better solution for this?

d-desiatkin replied:
We can always realize it ourselves with use of C struct.

d-desiatkin committed:
[chore](text-overflow): Add a reminder to add fallback.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[bugfix]: Crop the content area of fragments in addition to child
fragments

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[chore](text-overflow): Delete comented code

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): Transfer Fragments placement to stacking
context. On current stage bugs still present

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): Extend functionality to propperly handle
text-ellipsis left and right. Now fragments passed to stacking context
stage, but I still need to propperly position them.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](text-overflow): Modify stacking context to support ellipsis
left/right start/end values

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[chore](text-overflow): Clean up code

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[chore](text-overflow): Add check to fragment truncation

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[chore](text-overflow): Debugging text_fragment truncation

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>
